egs2/librispeech_100/asr1_biasing/asr.sh

```diff
@@ -0,0 +1,1857 @@
+        fi
+    fi
+    if "${use_ngram}"; then
+         _opts+="--ngram_file ${ngram_exp}/${inference_ngram}"
+    fi
+
+    # 2. Generate run.sh
+    log "Generate '${asr_exp}/${inference_tag}/run.sh'. You can resume the process from stage 12 using this script"
+    mkdir -p "${asr_exp}/${inference_tag}"; echo "${run_args} --stage 12 \"\$@\"; exit \$?" > "${asr_exp}/${inference_tag}/run.sh"; chmod +x "${asr_exp}/${inference_tag}/run.sh"
+
+    inference_bin_tag=""
+    if [ ${asr_task} == "asr" ]; then
+        if "${use_k2}"; then
+            # Now only _nj=1 is verified if using k2
+            inference_bin_tag="_k2"
+
+            _opts+="--is_ctc_decoding ${k2_ctc_decoding} "
+            _opts+="--use_nbest_rescoring ${use_nbest_rescoring} "
+            _opts+="--num_paths ${num_paths} "
+            _opts+="--nll_batch_size ${nll_batch_size} "
+            _opts+="--k2_config ${k2_config} "
+        elif "${use_streaming}"; then
+            inference_bin_tag="_streaming"
+        elif "${use_maskctc}"; then
+            inference_bin_tag="_maskctc"
+        fi
+    fi
+
+    if "${eval_valid_set}"; then
+        _dsets="org/${valid_set} ${test_sets}"
+    else
+        _dsets="${test_sets}"
+    fi
+    for dset in ${_dsets}; do
+        _data="${data_feats}/${dset}"
+        _dir="${asr_exp}/${inference_tag}/${dset}"
+        _logdir="${_dir}/logdir"
+        mkdir -p "${_logdir}"
+
+        if "${biasing}"; then
+            python local/get_perutt_blist.py data/${dset}
+            biasing_opts="--perutt_blist data/${dset}/perutt_blist.json"
+        fi
+
+        _feats_type="$(<${_data}/feats_type)"
+        _audio_format="$(cat ${_data}/audio_format 2>/dev/null || echo ${audio_format})"
+        if [ "${_feats_type}" = raw ]; then
+            _scp=wav.scp
+            if [[ "${audio_format}" == *ark* ]]; then
+                _type=kaldi_ark
+            elif [[ "${_audio_format}" == *multi* ]]; then
+                _type=multi_columns_sound
+            else
+                _type=sound
+            fi
+        else
+            _scp=feats.scp
+            _type=kaldi_ark
+        fi
+
+        # 1. Split the key file
+        key_file=${_data}/${_scp}
+        split_scps=""
+        if "${use_k2}"; then
+          # Now only _nj=1 is verified if using k2
+          _nj=1
+        else
+          _nj=$(min "${inference_nj}" "$(<${key_file} wc -l)")
+        fi
+
+        for n in $(seq "${_nj}"); do
+            split_scps+=" ${_logdir}/keys.${n}.scp"
+        done
+        # shellcheck disable=SC2086
+        utils/split_scp.pl "${key_file}" ${split_scps}
+
+        # 2. Submit decoding jobs
+        log "Decoding started... log: '${_logdir}/asr_inference.*.log'"
+        rm -f "${_logdir}/*.log"
+        # shellcheck disable=SC2046,SC2086
+        ${_cmd} --gpu "${_ngpu}" JOB=1:"${_nj}" "${_logdir}"/asr_inference.JOB.log \
+            ${python} -m espnet2.bin.${asr_task}_inference${inference_bin_tag} \
```

Comment:
Probably you can add biasing-related decoding file above. Then you don't need the separate `decode.sh`.

Reply:
I will remove it now. The run.sh now can run through the entire pipeline (I am still working on the calculating biasing word error part and will add that in to `asr.sh` as well when it is done.)